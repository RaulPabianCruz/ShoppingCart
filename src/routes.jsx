import Container from './components/Container/Container';
import Home from './components/Home/Home';
import Shop from './components/Shop/Shop';
import Cart from './components/Cart/Cart';
import ErrorPage from './ErrorPage';

const routes = [
  {
    path: '/',
    element: <Container />,
    errorElement: <ErrorPage />,
    children: [
      { index: true, element: <Home /> },
      {
        path: 'shop',
        element: <Shop />,
      },
      {
        path: 'cart',
        element: <Cart />,
      },
    ],
  },
];

export default routes;
