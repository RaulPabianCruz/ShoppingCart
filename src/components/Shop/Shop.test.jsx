import { render, screen } from '@testing-library/react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import routes from '../../routes';

describe('Shop Component', () => {
  describe('Shop content', () => {
    it('matches snapshot', () => {
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });

      const { container } = render(<RouterProvider router={router} />);

      expect(container).toMatchSnapshot();
    });

    it('renders all 30 products', () => {
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });
      render(<RouterProvider router={router} />);

      const productElements = screen.getAllByTestId('product-card');

      expect(productElements.length).toBe(30);
    });
  });
});

describe('Card Component', () => {
  describe('Card content', () => {
    it('renders all card component info', () => {
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });
      render(<RouterProvider router={router} />);

      const productImages = screen.getAllByTestId('product-card-img');
      const productTitles = screen.getAllByTestId('product-card-title');
      const productDescriptions = screen.getAllByTestId('product-card-desc');
      const productPrices = screen.getAllByTestId('product-card-price');

      expect(productImages.length).toBe(30);
      expect(productTitles.length).toBe(30);
      expect(productDescriptions.length).toBe(30);
      expect(productPrices.length).toBe(30);
    });

    it("renders number input and 'Add to Cart' button", () => {
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });
      render(<RouterProvider router={router} />);

      const productQuantities = screen.getAllByRole('input', { name: '1' });
      const addButtons = screen.getAllByRole('button', { name: 'Add to Cart' });

      expect(productQuantities.length).toBe(30);
      expect(addButtons.length).toBe(30);
    });
  });

  describe('Card Functionality', () => {
    it('updates item count display when an item is added', async () => {
      const user = userEvent.setup();
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });
      render(<RouterProvider router={router} />);

      const addButtons = screen.getAllByRole('button', { name: 'Add to Cart' });
      await user.click(addButtons[0]);

      expect(screen.getByTestId('num-of-items').textContent).toBe('1');
    });

    it('updates item count display based on quantity of item added', async () => {
      const user = userEvent.setup();
      const router = createMemoryRouter(routes, { initialEntries: ['/shop'] });
      render(<RouterProvider router={router} />);

      const productQuantities = screen.getAllByRole('input', { name: '1' });
      const addButtons = screen.getAllByRole('button', { name: 'Add to Cart' });
      await user.click(productQuantities[0]);
      await user.keyboard('[Backspace][Digit5]');
      await user.click(addButtons[0]);

      expect(screen.getByTestId('num-of-items').textContent).toBe('5');
    });
  });
});
