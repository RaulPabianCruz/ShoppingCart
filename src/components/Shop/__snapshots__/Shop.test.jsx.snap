// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Component > Shop content > matches snapshot 1`] = `
<div>
  <div
    class="_mainContainer_523845"
  >
    <header
      class="_header_523845"
    >
      <img
        alt="The e-Mart Logo"
        class="_logo_523845 _headerItem_523845"
        src="/src/assets/store.svg"
      />
      <h1
        class="_titleName_523845 _headerItem_523845"
      >
        The e-Mart!
      </h1>
      <a
        class="_links_523845 _headerItem_523845"
        href="/"
      >
        Home
      </a>
      <a
        class="_links_523845 _headerItem_523845"
        href="/shop"
      >
        Shop
      </a>
      <p
        class="_itemCount_523845 _headerItem_523845"
        data-testid="num-of-items"
      >
        0
      </p>
      <a
        class="_links_523845 _headerItem_523845 _iconLink_523845"
        data-testid="cart-link"
        href="/cart"
      >
        <img
          alt="Cart"
          class="_icon_523845"
          src="/src/assets/cart.svg"
        />
      </a>
    </header>
    <div>
      <h1>
        Placeholder Shop
      </h1>
    </div>
  </div>
</div>
`;
